In [38]:
import numpy as np
import pandas as pd
import seaborn as sb
import matplotlib.pyplot as plt
import sklearn

from pandas import Series, DataFrame
from pylab import rcParams
from sklearn import preprocessing
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn import metrics
from sklearn.metrics import classification_report
from statsmodels.stats.outliers_influence import variance_inflation_factor  #for VIF
from sklearn.ensemble import RandomForestClassifier

## Task 3
#### Import “Train.csv” into your Jupyter notebook. 

In [40]:
pdata = pd.read_csv('C:\\Users\\stunn\\Notebook\\Capstone\\Train.csv')

In [41]:
pdata.head()

ID  LIMIT_BAL  SEX  EDUCATION  MARRIAGE  AGE  PAY_0  PAY_2  PAY_3  PAY_4  \
0   2     120000    2          2         2   26     -1      2      0      0   
1   4      50000    2          2         1   37      0      0      0      0   
2   5      50000    1          2         1   57     -1      0     -1      0   
3   6      50000    1          1         2   37      0      0      0      0   
4   7     500000    1          1         2   29      0      0      0      0   

   ...  BILL_AMT4  BILL_AMT5  BILL_AMT6  PAY_AMT1  PAY_AMT2  PAY_AMT3  \
0  ...     3272.0     3455.0     3261.0       0.0    1000.0    1000.0   
1  ...    28314.0    28959.0    29547.0    2000.0    2019.0    1200.0   
2  ...    20940.0    19146.0    19131.0    2000.0   36681.0   10000.0   
3  ...    19394.0    19619.0    20024.0    2500.0    1815.0     657.0   
4  ...   542653.0   483003.0   473944.0   55000.0   40000.0   38000.0   

   PAY_AMT4  PAY_AMT5  PAY_AMT6  Default  
0    1000.0       0.0    2000.0        1  
1    1100.0    1069.0    1000.0        0  
2    9000.0     689.0     679.0        0  
3    1000.0    1000.0     800.0        0  
4   20239.0   13750.0   13770.0        0  

[5 rows x 25 columns]

In [42]:
# Check for missing values if any.
pdata.isnull().values.any()

False

In [43]:
# Convert the relevant variables such as payment variables (Pay0-Pay6 and customer related variables) to categorical variables as appropriate.

payment_columns = ['SEX','EDUCATION','MARRIAGE','PAY_0', 'PAY_2', 'PAY_3', 'PAY_4', 'PAY_5', 'PAY_6']
pdata[payment_columns] = pdata[payment_columns].astype('category')

In [44]:
#pdata.info()

In [45]:
# Dropping irrelevant columns

pdata.drop(['ID', 'BILL_AMT2', 'BILL_AMT3', 'BILL_AMT4', 'BILL_AMT5', 'BILL_AMT6'],axis=1,inplace=True)
pdata.head()

LIMIT_BAL SEX EDUCATION MARRIAGE  AGE PAY_0 PAY_2 PAY_3 PAY_4 PAY_5 PAY_6  \
0     120000   2         2        2   26    -1     2     0     0     0     2   
1      50000   2         2        1   37     0     0     0     0     0     0   
2      50000   1         2        1   57    -1     0    -1     0     0     0   
3      50000   1         1        2   37     0     0     0     0     0     0   
4     500000   1         1        2   29     0     0     0     0     0     0   

   BILL_AMT1  PAY_AMT1  PAY_AMT2  PAY_AMT3  PAY_AMT4  PAY_AMT5  PAY_AMT6  \
0       2682       0.0    1000.0    1000.0    1000.0       0.0    2000.0   
1      46990    2000.0    2019.0    1200.0    1100.0    1069.0    1000.0   
2       8617    2000.0   36681.0   10000.0    9000.0     689.0     679.0   
3      64400    2500.0    1815.0     657.0    1000.0    1000.0     800.0   
4     367965   55000.0   40000.0   38000.0   20239.0   13750.0   13770.0   

   Default  
0        1  
1        0  
2        0  
3        0  
4        0

Variable(s) that would cause multicollinearity: 'BILL_AMT1', 'BILL_AMT2', 'BILL_AMT3', 'BILL_AMT4', 'BILL_AMT5', 'BILL_AMT6'

In [47]:
X_train = pdata.iloc[:,0:-1].values
y_train = pdata['Default'].values

## Task 3.1	

#### Fit a random forest model on “Train.csv” with a random state of 1, 500 estimators, a maximum depth of 3 and a maximum feature of 3. 

In [49]:
rfcl = RandomForestClassifier(n_estimators = 500, max_depth=3, random_state=1, max_features=3)
rfcl = rfcl.fit(X_train, y_train)

In [50]:
pdataT = pd.read_csv('C:\\Users\\stunn\\Notebook\\Capstone\\Test.csv')
pdataT.head()

ID  LIMIT_BAL  SEX  EDUCATION  MARRIAGE  AGE  PAY_0  PAY_2  PAY_3  \
0   5170     140000    2          2         2   26      0      0      2   
1  22727      90000    2          1         2   31      1     -2     -2   
2   4884     150000    2          1         2   32      2      2      2   
3   7142      30000    2          2         2   21     -2     -2     -2   
4  19043      70000    2          2         2   23      2      2      2   

   PAY_4  ...  BILL_AMT4  BILL_AMT5  BILL_AMT6  PAY_AMT1  PAY_AMT2  PAY_AMT3  \
0      0  ...     125046    46364.0      83460    125000       6.0    4578.0   
1     -2  ...          0        0.0          0         0       0.0       0.0   
2      0  ...     123954   131543.0     134319      6000    3000.0    4600.0   
3     -2  ...       4311     1551.0          0       780    3260.0    4311.0   
4      0  ...      18013    19315.0      19859      2850       0.0    1400.0   

   PAY_AMT4  PAY_AMT5  PAY_AMT6  Default  
0    1427.0   87179.0   26291.0        1  
1       0.0       0.0       0.0        0  
2    9800.0    5100.0    5000.0        1  
3    1551.0       0.0       0.0        0  
4    1600.0    1000.0       0.0        1  

[5 rows x 25 columns]

In [51]:
# Convert the relevant variables such as payment variables (Pay0-Pay6 and customer related variables) to categorical variables as appropriate.

payment_columns = ['SEX','EDUCATION','MARRIAGE','PAY_0', 'PAY_2', 'PAY_3', 'PAY_4', 'PAY_5', 'PAY_6']
pdataT[payment_columns] = pdataT[payment_columns].astype('category')

In [52]:
pdataT.drop(['ID', 'BILL_AMT2', 'BILL_AMT3', 'BILL_AMT4', 'BILL_AMT5', 'BILL_AMT6'],axis=1,inplace=True)
pdataT.head()

LIMIT_BAL SEX EDUCATION MARRIAGE  AGE PAY_0 PAY_2 PAY_3 PAY_4 PAY_5 PAY_6  \
0     140000   2         2        2   26     0     0     2     0     0    -1   
1      90000   2         1        2   31     1    -2    -2    -2    -2    -2   
2     150000   2         1        2   32     2     2     2     0     0     2   
3      30000   2         2        2   21    -2    -2    -2    -2    -2    -2   
4      70000   2         2        2   23     2     2     2     0     0     2   

   BILL_AMT1  PAY_AMT1  PAY_AMT2  PAY_AMT3  PAY_AMT4  PAY_AMT5  PAY_AMT6  \
0    28282.0    125000       6.0    4578.0    1427.0   87179.0   26291.0   
1        0.0         0       0.0       0.0       0.0       0.0       0.0   
2   120800.0      6000    3000.0    4600.0    9800.0    5100.0    5000.0   
3        0.0       780    3260.0    4311.0    1551.0       0.0       0.0   
4    14871.0      2850       0.0    1400.0    1600.0    1000.0       0.0   

   Default  
0        1  
1        0  
2        1  
3        0  
4        1

In [53]:
X_test = pdataT.iloc[:,:-1].values
y_test = pdataT['Default'].values

In [54]:
y_pred = rfcl.predict(X_test) 

## Task 3.2	

#### Evaluate the confusion matrix, F1 scores and accuracy. Compare the random forest model with the logistic regression from Task 2. State your observations in a markdown cell in your Jupyter notebook.

In [56]:
from sklearn.metrics import confusion_matrix
confusion_matrix = confusion_matrix(y_test, y_pred)
confusion_matrix

array([[3013,  105],
       [ 672,  210]], dtype=int64)

In [57]:
print(metrics.classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.82      0.97      0.89      3118
           1       0.67      0.24      0.35       882

    accuracy                           0.81      4000
   macro avg       0.74      0.60      0.62      4000
weighted avg       0.78      0.81      0.77      4000



Observations:

Precision: precision for 1 improved from 38% to 67%

Recall: recall for 1 worsen from 65% to 24%

f1 score: F1 for 0 improved but got worse for 1

Accuracy: Improved to 81 % (from 70%)

*standardscaler not used